rasterio/transform.py

```diff
@@ -62,48 +63,72 @@ def array_bounds(height, width, transform):
     return w, s, e, n
 
 
-def xy(row, col, transform, offset='center'):
-    """Returns the coordinates ``(x, y)`` of a pixel at `row` and `col`.
+def xy(transform, rows, cols, offset='center'):
+    """Returns the x and y coordinates of pixels at `rows` and `cols`.
     The pixel's center is returned by default, but a corner can be returned
     by setting `offset` to one of `ul, ur, ll, lr`.
 
     Parameters
     ----------
-    row : int
-        Pixel row.
-    col : int
-        Pixel column.
     transform : affine.Affine
         Transformation from pixel coordinates to coordinate reference system.
+    rows : list or int
+        Pixel rows.
+    cols : list or int
+        Pixel columns.
     offset : str, optional
         Determines if the returned coordinates are for the center of the
         pixel or for a corner.
 
     Returns
     -------
-    tuple
-        ``(x, y)``
+    xs : list
+        x coordinates in coordinate reference system
+    ys : list
+        y coordinates in coordinate reference system
     """
 
-    cr = (col, row)
+    single_col = False
+    single_row = False
+    if not isinstance(cols, collections.Iterable):
+        cols = [cols]
+        single_col = True
+    if not isinstance(rows, collections.Iterable):
+        rows = [rows]
+        single_row = True
+
+    xs = []
+    ys = []
+    for col, row in zip(cols, rows):
+        cr = (col, row)
+
+        if offset == 'center':
+            x, y = transform * transform.translation(0.5, 0.5) * cr
+        elif offset == 'ul':
+            x, y = transform * cr
+        elif offset == 'ur':
+            x, y = transform * transform.translation(1, 0) * cr
+        elif offset == 'll':
+            x, y = transform * transform.translation(0, 1) * cr
+        elif offset == 'lr':
+            x, y = transform * transform.translation(1, 1) * cr
+        else:
```

Comment:
In the code above, all the operations should be `transform.translation(offset) * transform * cr` because that's how affine matrices associate mathematically.